sm/sha3/sha3.c

```diff
@@ -120,16 +122,27 @@ int sha3_update(sha3_ctx_t *c, const void *data, size_t len)
     size_t i;
     int j;
 
+    // wait until all events have been handled.
+    // prevent deadlock by consuming incoming IPIs.
+    uint32_t incoming_ipi = 0;
+
     j = c->pt;
     for (i = 0; i < len; i++) {
         c->st.b[j++] ^= ((const uint8_t *) data)[i];
         if (j >= c->rsiz) {
             sha3_keccakf(c->st.q);
             j = 0;
+            incoming_ipi |= atomic_swap(HLS()->ipi, 0);
```

Comment:
We need some more explanation in the comments for this one. It 'shouldn't' deadlock ever, but we've obviously been seeing strange behaviors.
At least document that and mark as TODO needing more investigation.

Reply:
I will def add another TODO there.